### PR TITLE
Add support for querying code interpreter

### LINKED ISSUE #60: Computational Q&A
# Describe the issue

Add computational Q&A capabilities to RAG pipeline. The pipeline should decide when computation/similarity search is necessary to run.

### PR DESCRIPTION
**Pull request by homanp**
# What does this PR do?

Adds support for running code interpreter on structured docs. 

Fixes #60

### CODE REVIEW

**Review by ValentaTomas — COMMENTED** [review]
I recommend updating to e2b SDK `0.14.7`.

File: service/code_interpreter.py [hunk 1]
```diff
@@ -73,36 +82,65 @@ async def __aexit__(self, _exc_type, _exc_value, _traceback):
...
         """
-
         # TODO: Add support for xslx, json
-        files_code = "\n".join(
-            f'df{i} = pd.read_csv("{self._get_file_path(url)}") # {url}'
-            for i, url in enumerate(self.file_urls)
+        df = pd.read_csv(self.file_urls[0])
+        return df, self.file_urls[0]
+
+    def genreate_prompt(self, query: str) -> str:
```
Comment (ValentaTomas) [inline service/code_interpreter.py:94]: Typo in `generate_prompt`
```suggestion
    def generate_prompt(self, query: str) -> str:
```

File: service/code_interpreter.py [hunk 2]
```diff
@@ -73,36 +82,65 @@ async def __aexit__(self, _exc_type, _exc_value, _traceback):
...
+        if matches:
+            return matches[0].strip()
+        return ""
 
-"""
+    async def generate_code(
+        self,
+        query: str,
+    ) -> str:
+        content = self.genreate_prompt(query=query)
```
Comment (ValentaTomas) [inline service/code_interpreter.py:127]:
```suggestion
        content = self.generate_prompt(query=query)
```

File: service/code_interpreter.py [hunk 3]
```diff
@@ -73,36 +82,65 @@ async def __aexit__(self, _exc_type, _exc_value, _traceback):
...
         self.sandbox.close()
 
-    def get_files_code(self):
+    def get_dataframe(self):
         """
         Get the code to read the files in the sandbox.
         This can be used for instructing the LLM how to access the loaded files.
         """
-
         # TODO: Add support for xslx, json
```
Comment (ValentaTomas) [inline service/code_interpreter.py:90]: Right now this works only for one file, right?
```suggestion
        # TODO: Add support for multiple dataframes
```

File: service/code_interpreter.py [hunk 4]
```diff
@@ -73,36 +82,65 @@ async def __aexit__(self, _exc_type, _exc_value, _traceback):
...
+        {df.info()}
+
+        Using the provided dataframe (`df`), update the following python code using pandas that returns the answer to question: \"{query}\"
+        
+        This is the initial python code to be updated:
+        
+        ```python
+        import pandas as pd
+
+        df = pd.read_csv("{url}")
```
Comment (ValentaTomas) [inline service/code_interpreter.py:108]: I was thinking about this and maybe putting the public urls that can be used for downloading the data in the prompt may not be ok later on.
We can circumvent this by referring to the data with a different name, I just wanted to point this out from a privacy/security perspective.

**Comment by ValentaTomas** [comment]
I improved the sandbox lifecycle handling with [a0a1cf3](https://github.com/superagent-ai/super-rag/pull/66/commits/a0a1cf370a671fdd8fb9d5148b37ca176e39aaae) to ensure the sandbox is not left hanging where there is an exception (for example when downloading files).

File: service/code_interpreter.py (hunk 4 above)
Comment (homanp) [reply service/code_interpreter.py:115]: Correct, would you de-indentify on your end?

**Comment by homanp** [comment]
Thank you for support @ValentaTomas ❤️

**Pull request merged**